test/puppetlabs/puppetdb/integration/smoke.clj

```diff
@@ -27,3 +28,94 @@
+     ; {:name "user_defined_object" :code "Car(abc123)" :expected "Car({'regnbr' => 'abc123'})"}
+     ])
+
+  (deftest ^:integration rich-data-agent-run
+    (with-open [pg (int/setup-postgres)
+                pdb (int/run-puppetdb pg {})
+                ps (int/run-puppet-server [pdb] {:agent {:rich-data true}})]
+
+      (testing "Agent run succeeds"
+        (let [manifest (str "type Car = Object[attributes => {regnbr => String}];\n"
+                            (apply str (for [{:keys [name code]} rich-data-tests]
+                                         (str "notify { '" name "':"
+                                              "  message => " code
+                                              "}"))))
+              {:keys [out]} (int/run-puppet-as "my-agent" ps pdb manifest {:rich-data true})]
+          (is (re-find #"a-z" out))))
+
+      (testing "Agent run data can be queried"
+        (are [query result] (= result (int/pql-query pdb query))
+          "nodes[certname] {}" [{:certname "my-agent"}]
+          "catalogs[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "factsets[certname, environment] {}" [{:certname "my-agent", :environment "production"}]
+          "reports[certname, environment] {}" [{:certname "my-agent", :environment "production"}]))
+
+      (testing "Catalog resource parameters stored in proper format"
+        (let [[{{all-resources :data} :resources} :as result] (int/pql-query pdb "catalogs[certname, resources] { resources { type = 'Notify' and title = 'regexp' } }")
+              filtered-resources (filter
+                                   #(not (or (= "Stage" (:type %))
+                                             (and (= "Class" (:type %))
+                                                  (or (= "main" (:title %))
+                                                      (= "Settings" (:title %))))))
+                                   all-resources)]
+          (is (= 1 (count result)))
+          (is (= "my-agent" (:certname (first result))))
+          (is (= (count rich-data-tests) (count filtered-resources)))
+
+          (doseq [resource filtered-resources]
+            (let [stored-value (:message (:parameters resource))
+                  filter-fn (fn [data] (= (:title resource) (:name data)))
+                  [test-resource :as filtered-resources] (filter filter-fn rich-data-tests)]
+              (is (= 1 (count filtered-resources)))
+
+              (if (:valid? test-resource)
+                (is ((:valid? test-resource) stored-value))
+                (is (= (:expected test-resource) stored-value)))))))
+
+      (testing "transaction-uuid"
+        (let [catalog-uuid (first (int/pql-query pdb "catalogs [transaction_uuid] {}"))
+              report-uuid (first (int/pql-query pdb "reports [transaction_uuid] {}"))]
+          (testing "is available from puppetdb"
+            (is (not (nil? catalog-uuid)))
+            (is (not (nil? report-uuid))))
```

Comment:
Are we specifically looking for `nil` if so, then this is just right, but if not, perhaps change this to just `(is (not catalog-uuid)), etc.

Reply:
this is another one taken from above, so I'm tempted to leave it as is